### PR TITLE
Documentation fixes

### PR DESCRIPTION
**Pull request by kdschlosser**
DO NOT MERGE THIS YET.

This will fix all of the documentation build errors. I have to double check that the CI changes are working properly.

This PR fixes the issue with the last release compiled docs being updated if a push happens to the master branch.

This PR also builds in a feature that will cause a PR CI to fail if something gets added/changed and it's not documented. only a maintainer of the repo will be able to bypass this but in order to do so they will need to invalidate the CI cache that holds the list of things that are currently not documented. If an API naming change takes place and nothing else the cache will also have to be invalidated. I didn't want to have the build fail because of current things that are not documented, Moving forward we want the build to fail if something is not documented. This is being done because I know there is going to be a code change coming up that is going to move large portions of the API to being private and since they are not public any more there is no need to worry about them being documented.

This can very easily be turned off of it turns out to be too much of a headache to use. I am sure it will need some tweaking to get 100% right.

### CODE REVIEW

File: src/misc/lv_style_gen.h
Comment (kisvegabor) [inline src/misc/lv_style_gen.h]: Why were these changed?

**Review by kisvegabor — COMMENTED** [review]
Awesome! :tada: 

There are a few remaining issues in the [docs](https://github.com/lvgl/lvgl/actions/runs/10034527798/job/27729185459?pr=6542#step:7:29).


I can't see why [these](https://github.com/lvgl/lvgl/actions/runs/10034527801/job/27729186353?pr=6542#step:7:363) are happening.

Why aren't there missing docs error for the functions in [`lv_style_gen.h`](https://github.com/lvgl/lvgl/pull/6542/files#diff-1a5d91c385e4dbc4203dcf821ea5733807a4b9b2248a2269ec8d7dceadcdd284)?

Could you add a `README.md` (similar to the [tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md)) file in the docs folder to explain:
- How the docs are built
- How to build them offline
- A basic guide to `rst`. A short example code can work too where the key features are demonstrated.

File: src/misc/lv_style_gen.h
Comment (kdschlosser) [reply src/misc/lv_style_gen.h]: Those were changed because of doxygen having issues with the LV_ATTRIBUTE_EXTERN_DATA macro.

As I said in the PR comment, DO NOT MERGE THIS.. I said that because I need the CI build to run to make sure that there are no build errors which is obviously is that was caused by a change I had made. I have to dig through the code in order to figure out the way to pass the doxygen config parameter to process and expand the macros prior to collecting the documentation from the header files.

There is more work that needs to be done to get everything to build 100%, give me another day or two to get the last remaining bits sorted out.

Comment (kdschlosser) [reply src/misc/lv_style_gen.h]: as far as the missing docs errors. If you click on actions in GitHub and locate the build for the documentation and click on it. There is going to be an attached artifact. That artifact contains a snapshot of everything that is missing in terms of docstrings. I can set the build to fail for missing documentation but the issue there is so much of LVGL is currently undocumented the only person that would be able to document the missing bits is one of the maintainers of LVGL. It could possibly be that only yourself is going to be able to do it. The build will always fail of that documentation is not finished and instead of lumping that amount of work onto someone like yourself right away I opted to accept the current missing documentations and only have the build fail is something new has been added without a docstring or something has changed and the docstring has not been updated. In order to do that we needed to have a snapshot of the state of the documentation which gets saved as a cache and also set as an artifact for inspection. 


If you as a maintainer make the decision to allow a PR to be merged even tho the doc build is failing because there is a missing docstring you can go into the actions and locate the build and invalidate the cache. The nxt time the documentation gets built the build will pass and the new information on the missing docstrings will then be added to the cached files.

**Comment by kdschlosser** [comment]
@kisvegabor 

I need you to go into github actions and under the build docs job can you please invalidate the cache.

**Comment by kisvegabor** [comment]
It seems only `MISSING-DOCSTRINGS-v1` was related. I've removed it.

File: src/misc/lv_style_gen.h
Comment (kisvegabor) [reply src/misc/lv_style_gen.h]: Let's separate this into 2 things:
1. There are no `Sphinx` errors: it seems we are pretty close.
2. There are no `Doxygen` error: we are working on it with Vic. (It will take some time for sure)

**Comment by kdschlosser** [comment]
OK the documentation actually did build successfully. It wasn't until it was cleaning up that there was a problem. That's easy to fix and I will take care of it shortly. 

Now it's up to you to decide if you want to merge this PR at this point. Anything that exists that is not currently documented will be stored in that cache you just cleared and there is also a build artifact that is created that contains that information as well. If something gets added or the API changes in any way and the documentation has not been updated properly then the build will fail until the documentation gets added/corrected.

This can be bypassed simply by clearing that cache like you just did and running the build again. This will allow the build to complete successfully and a cache will be created again with the new undocumented items added to it. This ideally should not be done but it can be if the need should arise. It's not a huge deal to clear that cache and rerun the build if needed. This will help to keep things documented properly.

I am still messing about with sphinx to turn the build warnings and errors into something I can stop the build as a failure. I only want to do this at the end of the build so all of the errors show all at once and can be fixed in one go instead of having to build, fix, build, fix, build fix..... which is really annoying to have to do..

File: docs/README.md
Comment (kisvegabor) [inline docs/README.md]:
```suggestion
### Bulleted lists
```

Same for others

File: docs/README.md
Comment (kisvegabor) [inline docs/README.md:186]: These decorations shall be removed

File: docs/README.md
Comment (kisvegabor) [inline docs/README.md:71]:
```suggestion
## .rst cheat sheet
```

File: docs/README.md
Comment (kisvegabor) [inline docs/README.md:257]: Please shortly describe the internal build process, and highlight the custom non trivial parts.

**Review by kisvegabor — COMMENTED** [review]
Just to be sure I understand correctly: the missing API, enum filed, etc docs are still producing warnings but the currently existing one are cached and silenced, right?

File: docs/README.md
Comment (kdschlosser) [reply docs/README.md:71]: It is set as for developers because doxygen syntax going to be added to that section as well.

File: docs/README.md
Comment (kdschlosser) [reply docs/README.md:186]: ???

File: docs/README.md
Comment (kdschlosser) [reply docs/README.md:257]: You want me to write something up for the internal build process of the documentation? Do you have any idea of how much that is? It would be a literal book, as in several hundred pages long. Doing a summation of what it does is only going to leave a person wondering what in the hell I am trying to convey. Best to not explain the actual process and only give the information that needs to be given and nothing more. There is no need to explain how the build system works. All they need to know is what is allowed and what is not allowed. I don't believe I am able to highlight sections of text that is inside of a code block. I know I can do that with sphinx but I do not believe that is something that is able to be done using markdown.

File: src/misc/lv_style_gen.h
Comment (kdschlosser) [reply src/misc/lv_style_gen.h]: most of the doxygen errors are suppressed and that is what Vic is working on. I have all but a single error/warning sorted out with sphinx. The one error is because there are 2 lvgl.h files and it is pitching a fit about it. I don't know why there are 2 but there is also one that is located in the src directory as well as in the root of the repo.

I managed to get the `LV_ATTRIBUTE_EXTERN_DATA` issue all sorted out so that is now working properly. I did have to update the style gen script to change the heading primer. 

It is building with only that single error about there being more than one lvgl.h file and I have tried every which way I can think of to deal with that and nothing seems to be working. If the file gets deleted the docs will build error and warning free...

**Comment by kdschlosser** [comment]
@kisvegabor 

I am going to need you to clear that cache again and restart the documentation build if it fails on this build.

If it doesn't fail then this PR is able to be merged how it is.

You are going to also want to go into the 9.1 release branch and in the actions for that branch restart the build for the documentation. That should update the documentation so it will not have the stuff from the master branch in it anymore. IDK if I told you that I found and corrected that issue.

File: docs/README.md
Comment (kisvegabor) [reply docs/README.md:257]: Don't need to go into the very detail. I mean e.g. We use Doxygen, breath converts it to Sphinx, where the relevant scripts are, link and a few words about the CI, etc. 

The goal is to move this information out of your head to the public a little 😄

**Comment by kisvegabor** [comment]
>  IDK if I told you that I found and corrected that issue.

Awesome! Thanks.

---

It passes now. :tada: Please fix the conflict and I can merge it.

**Comment by kdschlosser** [comment]
It is done!!!

File: docs/README.md
Comment (kdschlosser) [reply docs/README.md:257]: NO NO NO. If I do that then I have no job security... LOL

**Comment by kdschlosser** [comment]
@kisvegabor 

Did you take a look at the artifacts for the documentation build? In there are all the things that are not documented current in the lvgl header files. I want to make sure you know there is a list that does get output so we have something to work off of when adding the documentation. This way we don't have to read each and every single header file. Most of the entries in that artifact file have the file name and line number. The entries that don't it is easy enough to do a search using the name that is given.

**Comment by kdschlosser** [comment]
@kisvegabor 

The docs are failing because the cache needs to be reset. the PR's that were merged into this PR had new stuff that was not documented in it. 


```
 MissingFunctionReturnDoc: lv_array_back
    FILE: /home/runner/work/lvgl/lvgl/src/misc/lv_array.h
    LINE: 199

MissingMacroDoc: LV_ARRAY_DEFAULT_SHRINK_RATIO
    FILE: /home/runner/work/lvgl/lvgl/src/misc/lv_array.h
    LINE: 27
```

```c
/**
 * Returns a pointer to the last element in the array.
 * @param array pointer to an `lv_array_t` variable
 */
static inline void * lv_array_back(const lv_array_t * array)
{
    return lv_array_at(array, lv_array_size(array) - 1);
}
```

So it is  now confirmed that it is working properly.. YAY!!!

**Comment by kisvegabor** [comment]
Great! Re-running now.

**Review by kisvegabor — DISMISSED** [review]

**Comment by lhdjply** [comment]
The build can still pass CI even with error.

**Comment by kisvegabor** [comment]
@lhdjply Nice catch!

**Comment by kdschlosser** [comment]
working on it. Will get it fixed in a minute. I also have to get rid of that other lvgl.h file too.

File: docs/build.py
Comment (becseya) [inline docs/build.py:402]: I don't really get this part. At what point does `build.py` gets launched multiple times?

File: docs/build.py
Comment (becseya) [inline docs/build.py:204]: I'll pin my comment to this line, but it's a more generic feedback.

I feel like this script is getting a bit messy and hard to understand for future contributors. E.g.:
- function definitions and instructions mixing in the execution flow
- high level nesting and code repetition
- the idea of a script reading its own stdout / stderr feels a awkward

My suggestions would be:
- use `os.walk()`
- move class and method definitions to the beginning of file / separate module
- move "clean", "doxy", "pdf" and "html" phases to standalone methods which can be called or not depending on the cli arguments
- I'd expect a flag to exist which makes doxygen/sphinx to return an error code upon any warning. We should pass that, and watch the return value instead of monitoring the output. If that's not possible for some reason, we could still use an external script - bash might be a better fit for this role -  to catch keywords in std stream.

I'm happy to make these changes in an upcoming PR if you want to move forward with this one

File: docs/build.py
Comment (kdschlosser) [reply docs/build.py:402]: It's how multiprocessing works when using python. The script gets run more than once because sphinx is being instructed to use multiple jobs which is what launches it more than once.

File: docs/build.py
Comment (kdschlosser) [reply docs/build.py:204]: If you believe you are able to write a better script then have at it. Now remember this script is not just used to build the documentation so make sure that everything else still functions properly with the script you write.

File: docs/build.py
Comment (becseya) [reply docs/build.py:402]: I've added a line which generates a randomly named file upon the execution of `run()`, removed the `if __name__ == '__main__':` guard and called `build.py`. Only a single file was generated.

Could you explain how to reproduce the "script being run more than once" scenario, please?

Comment (kdschlosser) [reply docs/build.py:402]: Of course it is only going to make a single file upon execution of run. That's what it is supposed to do. Now if you make the file at the module level you will find that more than one file gets made. You can also just add a print statement at the module level as well so each time the thing executes there will be something printed out.

Comment (kdschlosser) [reply docs/build.py:402]: Also this is only going to hold true if you are using a platform other than Windows.

Comment (becseya) [reply docs/build.py:402]: I'm using Linux, and still unable to recreate any scenario resembling what we are discussing.

I see no problem with the `if __name__ == '__main__':` guard as it's a common Python practice (although only recommended for modules, not standalone scripts). However I'd remove the comment as it might be confusing and the behavior it describes is not confirmed.

Comment (kdschlosser) [reply docs/build.py:402]: read...

https://docs.python.org/3/library/multiprocessing.html

specifically this section

https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods

Now read this...

https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-j

Comment (kdschlosser) [reply docs/build.py:402]: delete the `if __name__ == '__main__':` line then run it and see what happens. It will run the entire document build multiple times in parallel, ultimately crashing as a result.

Comment (becseya) [reply docs/build.py:402]: I managed to get the following error on a virtual Windows machine:
Screenshot from 2024-07-31 18-44-51
But the program exited successfully.

The situation is clearer now, my suggestion for the wording:
```
# we are invoking sphinx in multiprocessor mode (-j flag)
# which can cause waring messages / a crash under Windows
# (https://stackoverflow.com/a/50782028)
```

**Comment by vwheeler63** [comment]
@kdschlosser I don't know if you noticed the above (conflicts getting in the way of merging), but thought I would give you a heads-up just in case it helps.

**Comment by vwheeler63** [comment]
Good evening, Gabor!  ( @kisvegabor )

I am at a point where I need to make a situation known to you:  The current `master` does not have the 201-ish source-code changes Kevin made to get `breathe` and `sphinx` to stop generating errors.  At least some of them were Doxygen warnings.  With the setup I have, I have a private branch where I also have Kevin's changes, and so the warning list I am working from **_does not_** contain the warnings Kevin already handled.  However, the files I make changes need to be a branch created directly off the latest `master` so as to minimize the impact when it comes time for me to submit.  I expect to have the whole set of changes (resulting from the 3181 Doxygen warnings examined) ready after 3 more working days.  Once I have the full set, I can split it up into different branches to reduce the file count per PR.  I don't know how many there will be total yet, so I will consult with you once I know the final number of the solution is not obvious (e.g. if there were only 20-40-ish more changed files, but based on what I have seen thus far, I am expecting more as of this writing).  If I submit those PRs before this PR (PR#6542) gets merged, I could cause more conflicts for it, and I don't want to do that.  I'm thinking the best thing I can do is make my upcoming changes on a branch off master, and then rebase that branch onto current `master` branch after this one gets merged, so that what I end up submitting has no conflicts and (hopefully) creates none for others.

I'm bringing this up to give you a "heads-up" that I am about 3 work days away from coming to a point where I am having to wait (on the merge of this PR), to submit my set of PRs, so I am starting to feel some urgency in getting this one merged.  Just wanted you to have the information so you can prioritize.

Over to you, kind sir!
--Vic

**Comment by kdschlosser** [comment]
@vwheeler63 

I was waiting for your changes to get merged and then I was going to solve any merge conflicts and get my fork/branch all inline with the master. Then I was going to dive into why errors are not being caught properly from sphinx. Once that has all been solved then I was going to dive into looking at how the doxygen/sphinx portion of the documentation came out and making any tweaks that may be necessary. 

Now that I know I am able to monkey patch breathe we have the ability to adjust the output of breathe to suit our needs.

**Comment by vwheeler63** [comment]
@kdschlosser That sounds exciting!  Am I understanding correctly that you want me to go ahead and apply my Doxygen-warning changes to the current `master` and then you will re-merge `master` into your `documentation-fixes` branch and sort it all out -- potential conflicts and all?

**Comment by kdschlosser** [comment]
whatever you need to do..


I came across another sphinx plugin that is like breathe,... This one has a neat ability, it reads comments that are formatted in sphinx ReST from the C header/source files. It will also read doxygen formatted docstrings as well but that portion of it is not 100% complete. I just started dinking about with it. It would be nice to phase out the doxygen syntax for ReST syntax due to it being quite a bit more feature rich.

**Comment by vwheeler63** [comment]
@kdschlosser  @kisvegabor 

> whatever you need to do..

Okay, I'm going for it.... (doing just the "is not documented" warnings against the current `master` branch).

I estimated 3 days based on an estimated number of changes at 10-20% of the 3000+ "is not documented" warnings really aren't documented, and so will not get my attention for this task.  If it is much higher than that, than it may extend a bit.  I've built an application that will put my editor in the target file and line number and show me the warning (and let me pick up where I left off in each new session) so as to streamline the workflow of this task, so I'll get back to you about what the % is turning out to be so everyone is informed about what to expect.

Kind regards,
Vic

**Comment by kisvegabor** [comment]
@vwheeler63 According to your experience so far, is there significant number of undocumented functions/members that you cannot complete on your own?

**Comment by vwheeler63** [comment]
> @vwheeler63 According to your experience so far, is there significant number of undocumented functions/members that you cannot complete on your own?

@kisvegabor Good question.  So far about 80% of the `"is not documented"` warnings had no documentation.  The vast majority of these were `struct`s.  There were probably 40-50 or so functions that had no documentation.  Here is an example from `lv_grid.h` starting at line 61:

```
void lv_grid_init(void);

void lv_obj_set_grid_dsc_array(lv_obj_t * obj, const int32_t col_dsc[], const int32_t row_dsc[]);

void lv_obj_set_grid_align(lv_obj_t * obj, lv_grid_align_t column_align, lv_grid_align_t row_align);
```

There were a few cases where arguments were different.  Based on my experience with LVGL v8.x there were a few where I was familiar with the data type which I could fill in descriptions for [1 or 2 from yesterday's batch]).  In other cases, I re-aligned the argument list with the function and marked the argument with "???" (following @kdschlosser 's pattern).  There were 5 of these total.  Here is an example:

```
/**
 * Initialize the draw buffer with given handlers.
 *
 * @param handlers             the draw buffer handlers to set
 * @param buf_malloc_cb        the callback to allocate memory for the buffer
 * @param buf_free_cb          the callback to free memory of the buffer
 * @param align_pointer_cb     the callback to align the buffer
 * @param invalidate_cache_cb  the callback to invalidate the cache of the buffer
 * @param flush_cache_cb       ???
 * @param width_to_stride_cb   the callback to calculate the stride based on the width and color format
 */
void lv_draw_buf_handlers_init(lv_draw_buf_handlers_t * handlers,
                               lv_draw_buf_malloc_cb buf_malloc_cb,
                               lv_draw_buf_free_cb buf_free_cb,
                               lv_draw_buf_align_cb align_pointer_cb,
                               lv_draw_buf_cache_operation_cb invalidate_cache_cb,
                               lv_draw_buf_cache_operation_cb flush_cache_cb,
                               lv_draw_buf_width_to_stride_cb width_to_stride_cb);
```

And in one weird case, it was FULLY documented, but none of it was being accepted by Doxygen:  ./src/others/vg_lite.h.  It is now fully Doxygen-ized and its contents and all the original author's documentation will make it into the end result.  I checked in [https://docs.lvgl.io/master/API/others/vg_lite_tvg/vg_lite.html](https://docs.lvgl.io/master/API/others/vg_lite_tvg/vg_lite.html) and all the members appear to be present, but all the comments (the doc-string parts) are (currently) missing.  That will be remedied once the version of `vg_lite.h` I have here makes it into `master`.  :-)

Kind regards,
Vic

**Comment by kdschlosser** [comment]
I wouldn't worry about any header files that do not start with `lv_` as they are not apart of the public API. If you can alter them so no warnings get emitted by doxygen that would be good enough. the docstrings located in those files do not going to get added to the LVGL documentation.

**Comment by vwheeler63** [comment]
> I wouldn't worry about any header files that do not start with `lv_` as they are not apart of the public API.

@kdschlosser  Understood.  Thanks.

**Comment by vwheeler63** [comment]
@kisvegabor @kdschlosser 

Hi, Gabor!

Well, I'm a day ahead of schedule -- I completed the list.  All existing documentation that Doxygen was "tripping" on that was not formatted correctly is now formatted correctly.  Total number of files involved:  75.  Given the additional hassle of getting 2 approvals, would you consider them all in one PR?  Or would you like me to split it into 2?

Kind regards,
Vic

**Comment by kdschlosser** [comment]
@vwheeler63 

What else you have in the way of pending PR's and other work that is still being done?? I just merged the master branch into this PR and resolved the merge conflicts.  It is more than likely going to fail the build because there is still some stuff I have to sort out but I do not want to do that until you are done doing what you are doing with the documentation.

File: docs/build.py
Comment (kdschlosser) [reply docs/build.py:402]: This explains what is happening just fine.  There is no need to change it. It explains in detail what is taking place. 

```python
# because sphinx can use multiple CPU cores when it compiles this requires
# launching the script multiple times from different subprocesses. We only want
# the build script to run a single time and that is when the user runs the
# script. By checking the package name of this file to see if it is the "main"
# script puts in place a block so the build script only runs a single time
# even when the script gets run from multiple processes.
```

The description you are using is actually incorrect. `-j` is used to specify the number of processes (jobs) to launch, it has nothing to do with the number of processors. I can set that number to 30 if I wanted to and the computer could only have 6 cores (processors) and it would compile just fine. There would not be any real benefit from doing that performance wise. It would add a lot of overhead with no return on that investment. I have coded it so that the number of processes that is being run matches the logical processor count on the machine.

**Comment by vwheeler63** [comment]
Hi, Kevin ( @kdschlosser )

> What else you have in the way of pending PR's

Just our mutual work on LVGL issue #6480 at the moment.  The 2 pending PRs are PR #6652 and PR #6653.  The final bit is an extensive overhaul in `lv_conf_template.h` (and the commensurate files that go along with that particular one) which is heavily documented, but most of it was not yet in Doxygen format yet.  That one turned out to be significant enough that I withdrew it from its original PR (one of the 2 above) and am making it its own because I believe Gabor will want to scrutinize it closely, since it is so heavily used by programmers who are _users_ of LVGL (as opposed to maintainers).  I also fixed a number of grammatical errors in it for clarity, which makes it quite a bit easier and nicer to use from the perspective of the LVGL _consumer_.

I suspect once those 2 PRs are merged (the first has 2 approvals already), that you will likely find very few conflicts EXCEPT where you needed to modify parameters to squelch `breathe/sphinx` warnings.  In some cases, I expect they will be identical (except perhaps for whitespace differences), and in a few cases, I was able to do the homework and find out the missing description (rather than a "???" placeholder) in which case I added it.  XuNeo helped in one of those cases because he knew what they were.

And finally, as I promised Gabor, I have been making a list of things I saw which I think need addressing.  One seemed serious enough to register as an issue #6659.  It was (perhaps?) by unusual fortune that I happened to have my editor open to _that_ portion of the `Kconfig` file while I was studying about the Linux `conf` program, used to set up Linux source code for a custom build of the kernel.  In the case of what I was looking at, I got to see some of its (compiler/parsing) front-end source code, which was clearly incompatible with the `Kconfig` file I was looking at!  But it is from a niche Linux world (NVIDIA), which I'm sure has its own Linux environment(s), as does IBM.  So the modern Linux `conf` program may have been upgraded to accept leading spaces in the file -- this I don't know (yet).

I also left a personal question for you on our forum chat.

--Vic

**Comment by vwheeler63** [comment]
> I wouldn't worry about any header files that do not start with `lv_` as they are not apart of the public API. If you can alter them so no warnings get emitted by doxygen that would be good enough. the docstrings located in those files are not going to be added to the LVGL documentation.

@kdschlosser  Have you considered the `lvgl.h` file?  It will not pass the "lv_" criterion, but it does have Doxygen documentation in it.  Example:  `LV_VERSION_CHECK(x,y,z)` macro and the in-line functions therein should be documented and available to LVGL application developers.  Perhaps a better filename criterion is `starts with "lv"` instead of "lv_"?

**Comment by kdschlosser** [comment]
in the build script I am adding that file separately because it is not in the `src` folder. I am supplying specific folders to the sphinx portion of the build to search for header files to add. I am not limiting that in doxygen... If doxygen does a recursive iteration over all of the folders then you may run into a problem with it picking up header files in places like `examples` or `env_support`.. we don't want any of the header files in the documentation from those folders. That is the reason why on the sphinx side of things i specifically tell the build what folders to look for and then manually add in the `lv_conf.h` and `lvgl.h` files. One issue I am running into is there are 2 `lvgl.h` files. there is also on in the src directory and that one needs to be excluded from the doxygen build as well.

I would like to be able to use a generic type config file for doxygen and pass config parameters through the command line if the config parameter includes any kind of a path statement that is not able to be wildcarded.

Currently what is being done is this.. in the config file there are entries that look like this...

```
EXCLUDE                = *#*#EXCLUDE#*#*
```

The build script copies the config file to the temp directory and then it opens the file and replaces `*#*#EXCLUDE#*#*` with the paths  the things to be excluded. The paths used are absolute paths. I would rather not have to open the file and edit it in order for it to work. I am pretty sure that the config settings are able to be set using command line arguments instead.


so whatever it is you come up with for the config settings let me know what they are and I can code them into the build script to be past to doxygen using the command line instead.

**Comment by kdschlosser** [comment]
@kisvegabor 

There is a HUGE naming conflict that is happening now with the documentation. This is because all of the underscores were removed for the structure names which makes the structure name the same as the typedef name. breathe and sphinx do not know how to handle that because of the duplicate names. The underscores need to be added back to the structure names so the structure name is different from the typedef name.

**Comment by kisvegabor** [comment]
> The underscores need to be added back to the structure names so the structure name is different from the typedef name.

Ouch! Can you try if a simple search and replace of `typedef struct lv_` with `typedef struct _lv_` does the job?

**Comment by kdschlosser** [comment]
tried that. it doesn't

**Comment by kisvegabor** [comment]
See#https://github.com/lvgl/lvgl/pull/6688

**Comment by vwheeler63** [comment]
@kdschlosser @kisvegabor All checks are now passed....  What changed to make it work?  (Reason I ask is that I am not showing any changes to this branch in Kevin's LVGL repository, so it must have been another merge?)

**Comment by kisvegabor** [comment]
Only Kevin can tell :smile:

**Comment by kdschlosser** [comment]
@kisvegabor 

Do me a favor and run the CI again to see if it passes.

**Comment by kdschlosser** [comment]
never mind on that. see here...

https://github.com/lvgl/lvgl/actions/runs/10355936072/job/28664678567?pr=6542

**Comment by kisvegabor** [comment]
Let's merge this to fix the rst-t files ASAP to avoid the conflict hell. 

There are massive changes in the `py` files that are hard to follow for me. Finally have you replaced `breath` with a custom script? 

Let's keep the non-rst changes as minimal as possible to fix only the rst issues first. If an issue cannot be fixed without a complex custom script let's leave it as it is for now and focus on the ones which can be fixed with the existing tooling. It will be huge step forward already.

**Comment by vwheeler63** [comment]
> the ones which can be fixed with the existing tooling. It will be huge step forward already.

This makes a lot of sense, Gabor -- one _complete-able_ step at a time, each step giving gain.  If Kevin is adept with interactive re-basing, it should be easy to preserve the valuable work he did prior to our Zoom call, while preserving the more extensive work into another branch.  (I am hoping Kevin likes this idea.)

**Comment by vwheeler63** [comment]
Hi, Kevin!  ( @kdschlosser )

cc:  @kisvegabor 

You did some really valuable work on .RST files in this PR that handled a large number of Sphinx warnings (outside of the sphere of those caused by Doxygen warnings).  Those changes are very valuable, and especially now that I'm doing some re-writing and re-organizing of the first 5-or-so top-level pages in the documentation:  I would very much like to be introducing my own changes in a context that starts out having no (or very few) Sphinx warnings.  I don't know the best way to approach this, but one way is for me to cherry pick them from your changes in your [`documentation-fixes`](https://github.com/kdschlosser/lvgl/tree/documentation-fixes) branch.  But it seems like those changes should be their own PR (and they are already herein), and be merged prior to my changes, since my changes would create a bunch of conflicts without these valuable changes in them.

Another way (perhaps better) is for Kevin to use interactive rebasing to split this branch into 2 different branches, preserving the .RST changes in this branch, and the other changes in another branch to be a future PR?  In theory, this would only take a few minutes.  (Kevin, let me know if it would help for me to help with that -- I'm now relatively adept with interactive rebasing to split a branch into 2 different "topic branches".)

What do you guys think?

Kind regards,
Vic

**Comment by vwheeler63** [comment]
@kisvegabor @kdschlosser 

The more I think about this, the more obvious it is that I should go ahead.  At the moment, cherry picking the .RST files from Kevin's branch is working well, so I will submit those first as its own PR because I need these changes to work from.

**Comment by kisvegabor** [comment]
Cherry-picking only the RST changes looks safe and straight-forward. Please go ahead with it.

**Comment by vwheeler63** [comment]
> Cherry-picking only the RST changes looks safe and straight-forward. Please go ahead with it.

Okay!  In progress...

**Pull request closed without merging**
